### ISSUE TITLE
Sockets Used Rises, Network Dies, Fixed by running tcpdump

### ISSUE THREAD

**Issue by privatwolke** [coreos/bugs#2572]
# Issue Report #

## Bug ##

I'm creating this issue based on [this ServerFault question](https://serverfault.com/questions/959833). Short summary:

We run a small Elasticsearch cluster based on CoreOS/Azure. This setup has been working almost entriely maintenance-free for over 1.5 years. As of March 11th (which is when 2023.5.0 was released to stable) we're seeing nodes entering into a strange state after running for a few days.

"Strange state" means that network communication from the affected node to the other two nodes (which are deployed in the same subnet) breaks down. The machine is still reable from different subnets/peered virtual networks and has (very spotty) Internet connection with most requests timing out.

From monitoring the node we could see that the "sockets used" metric reported by `/proc/net/sockstat` increases from about 300 on a health node to over 4k. We were unsuccessful in lowering this number through the termination of processes (we shutdown the Elasticsearch container and even the entire docker runtime as well as several other processes).

When trying to run tcpdump on the machine, the metric immediately dropped down to a low level and network connectvitiy was restored. We then ran tcpdump on another affected machine through gdb with breakpoints set on all syscalls. It turned out what actually "fixes" the issue is tcpdump through libpcap creating a packet capture socket and turning on promiscous mode.

![sockets used metric](https://i.stack.imgur.com/qFpaF.png)

We've been in contact with Azure about this who are as confused by this as we are. I have noticed that CoreOS 2023.5.0 included a kernel upgrade from 4.14.96 to 4.19.25 which seems to include lots of changes to hv_netvsc (the driver that the Azure/Hyper-V provided network interface uses).

For full details please check out: https://serverfault.com/questions/959833
I have also captured system stats: https://gist.github.com/privatwolke/e7e2e7eb0272787765f5d3726f37107c

### Container Linux Version ###

```
$ cat /etc/os-release
NAME="Container Linux by CoreOS"
ID=coreos
VERSION=2023.5.0
VERSION_ID=2023.5.0
BUILD_ID=2019-03-09-0138
PRETTY_NAME="Container Linux by CoreOS 2023.5.0 (Rhyolite)"
ANSI_COLOR="38;5;75"
HOME_URL="https://coreos.com/"
BUG_REPORT_URL="https://issues.coreos.com"
COREOS_BOARD="amd64-usr"
```

### Environment ###

> What hardware/cloud provider/hypervisor is being used to run Container Linux?

Azure (seen on both Standard_L4s and Standard_DS2_v2 machines)

### Expected Behavior ###
Network keeps running smoothly and without interruptions.

### Actual Behavior ###
"sockets used" metric rises and network becomes unusable

**Comment by josephtsalisbury** [comment]
Have you found a way to consistently reproduce this issue, or does it only happen sporadically?

**Comment by dimaslv** [comment]
We have similar problem of "sockets: used" leak in /proc/net/sockstat after switching from tc fq to tc mq as default default_qdisc on 4.19.30 + Centos 7.6.
But we don't have problems with connectivity and counters still grow after running tcpdump.
I assume that it could be connected with in-kernel tcp pacing feature as it is turned on by switching fq off.
And it is reproducible in our case.

Haven't found any other similar reports, so just FYI.

**Comment by josephtsalisbury** [comment]
@dimaslv Do you have a specific workload or script that can be used to reproduce the issue?  Being able to reproduce it would be helpful, so we can test out patches and various kernel versions.

**Comment by dimaslv** [comment]
Problem is reproducible, but unfortunately I can't provide you with application, because it is proprietary Java app.

**Comment by privatwolke** [comment]
We have been working with the folks at Microsoft that maintain the networking part of hyperv. They have supplied us with a patch that we've been testing successfully for the last 9 days. We're working on getting it applied upstream.

**Comment by lucab** [comment]
@privatwolke thanks for the followup! Once the fix lands in the upstream tree, having a commit reference for tracking purposes here would be appreciated.

**Comment by dcui** [comment]
> We have been working with the folks at Microsoft that maintain the networking part of hyperv. They have supplied us with a patch that we've been testing successfully for the last 9 days. We're working on getting it applied upstream.

@privatwolke : can you please share the patch at a place like https://gist.github.com/ ?  I'm curious about the patch. :-)

**Comment by privatwolke** [comment]
@dcui, sure: https://gist.github.com/privatwolke/e58046288c105069c00896c7d7e99bd5
@lucab, will add a reference here once the patch lands upstream

**Comment by privatwolke** [comment]
The patch has been submitted: https://lkml.org/lkml/2019/4/30/1118

**Comment by privatwolke** [comment]
It has been applied to master: https://github.com/torvalds/linux/commit/93aa4792c3908eac87ddd368ee0fe0564148232b#diff-3c639e695024f40f659dd05cf1d24d1f

**Comment by bgilbert** [comment]
[4.19 backport](https://www.spinics.net/lists/netdev/msg571611.html).

**Comment by bgilbert** [comment]
Queued for 4.19.47.

**Comment by bgilbert** [comment]
The patch landed in kernel 4.19.47, which will be included in Container Linux alpha 2163.0.0, due shortly.  Could you test that the upcoming alpha fixes the problem?

**Comment by bgilbert** [comment]
This should be fixed in next week's stable release.  Thanks for reporting!
